File: core/src/main/java/to/klay/wspro/core/api/game/field/PlayArea.java
/*******************************************************************************
 *  WSPro - Trading Card Game Simulator
 *  Copyright (C) 2020  Klayton Killough
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published
 *  by the Free Software Foundation, version 3 of the License.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

package to.klay.wspro.core.api.game.field;

import to.klay.wspro.core.api.game.GameRuntimeException;
import to.klay.wspro.core.api.game.player.GamePlayer;
import to.klay.wspro.core.game.formats.standard.zones.PlayZone;

import java.util.Collection;
import java.util.EnumSet;

/**
 * Interacts with the player area of a particular player.<br/>
 * Interfaces with {@link PlayZone}'s within this area
 */
public interface PlayArea {

    /**
     * Obtains the owner of this play area
     * @return owner of the play area
     */
    GamePlayer getOwner();

    /**
     * Returns a collection of the {@link PlayZone}'s within this area, with no respect to ordering.
     * @return a collection of Play Zones, no ordering is guaranteed
     */
    Collection<PlayZone> getAllPlayZones();


    /**
     * Returns a collection of the {@link PlayZone}'s defined by the {@link EnumSet}.
     * @return a list of Play Zones, no ordering is guaranteed
     */
    Collection<PlayZone> getPlayZones(EnumSet<Zones> ZonesSet);

    /**
     * Returns an an individual Play Zone contained within the Play area.
     * @param zone  - the {@link Zones} constant representing the name of the Zone.
     * @return  -A Play Zone, provided the zone exists inside the area
     * @throws GameRuntimeException - if the zone doesnt exist inside the play area
     */
    PlayZone getPlayZone(Zones zone) throws GameRuntimeException;

    /**
     * Gets the Corresponding Marker zone when passed a Stage Zone
     * @param stageZone - an existing PlayZone on stage
     * @return Play Zone, provided the zone exists inside the area
     * @throws GameRuntimeException - if the zone doesnt correlate
     */
    public PlayZone getCorrespondingMarkerZoneOnStage(PlayZone stageZone) throws GameRuntimeException;
    
    void setOwner(GamePlayer owner);
    
}

File: web-client/src/game/logic/field/PlayArea.ts
/*******************************************************************************
 *  WSPro - Trading Card Game Simulator
 *  Copyright (C) 2020  Klayton Killough
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published
 *  by the Free Software Foundation, version 3 of the License.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

import {ZoneName} from "./ZoneName";
import PlayZone from "./PlayZone";
import * as PIXI from "pixi.js";
import PlayAreaView from "../../view/field/PlayAreaView";
import Player from "../Player";


const {CLIMAX, MEMORY, CENTER_STAGE_LEFT, CENTER_STAGE_RIGHT, STOCK, BACK_STAGE_RIGHT, WAITING_ROOM, CENTER_STAGE_MIDDLE, LEVEL, HAND, DECK, CLOCK, BACK_STAGE_LEFT} = ZoneName;
export default class PlayArea extends PIXI.Container{

    private zonesMap: Map<ZoneName, PlayZone> = new  Map<ZoneName, PlayZone>();
    view: PlayAreaView
    owner: Player

    constructor(owner : Player) {
        super();
        this.owner = owner;
        this.initZones();
        this.view = new PlayAreaView(this);
    }

    private initZones() {
        this.zonesMap.set(MEMORY, new PlayZone(MEMORY, this.owner));
        this.zonesMap.set(DECK, new PlayZone(DECK, this.owner));
        this.zonesMap.set(WAITING_ROOM, new PlayZone(WAITING_ROOM, this.owner));
        this.zonesMap.set(CLIMAX, new PlayZone(CLIMAX, this.owner));
        this.zonesMap.set(LEVEL, new PlayZone(LEVEL, this.owner));
        this.zonesMap.set(HAND, new PlayZone(HAND, this.owner));
        this.zonesMap.set(STOCK, new PlayZone(STOCK, this.owner));
        this.zonesMap.set(CLOCK, new PlayZone(CLOCK, this.owner));
        this.zonesMap.set(CENTER_STAGE_RIGHT, new PlayZone(CENTER_STAGE_RIGHT, this.owner));
        this.zonesMap.set(CENTER_STAGE_MIDDLE, new PlayZone(CENTER_STAGE_MIDDLE, this.owner));
        this.zonesMap.set(CENTER_STAGE_LEFT, new PlayZone(CENTER_STAGE_LEFT, this.owner));
        this.zonesMap.set(BACK_STAGE_LEFT, new PlayZone(BACK_STAGE_LEFT, this.owner));
        this.zonesMap.set(BACK_STAGE_RIGHT, new PlayZone(BACK_STAGE_RIGHT, this.owner));
    }

    getZone(name: ZoneName): PlayZone{
        let zone = this.zonesMap.get(name);

        if (zone !== undefined){
            return zone;
        } else {
            console.error(name + " was undefined in the PlayArea.");
            // @ts-ignore - this should break the code if it happens.
            return undefined;
        }
    }


}